File: kafka-0.7.0/kafka-0.7.0-core/src/main/scala/kafka/consumer/ZookeeperTopicEventWatcher.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kafka.consumer

import org.apache.log4j.Logger
import scala.collection.JavaConversions._
import kafka.utils.{Utils, ZkUtils, ZKStringSerializer}
import org.I0Itec.zkclient.{IZkStateListener, IZkChildListener, ZkClient}
import org.apache.zookeeper.Watcher.Event.KeeperState

class ZookeeperTopicEventWatcher(val config:ConsumerConfig,
    val eventHandler: TopicEventHandler[String]) {

  private val logger = Logger.getLogger(getClass)

  val lock = new Object()

  private var zkClient: ZkClient = new ZkClient(config.zkConnect, config.zkSessionTimeoutMs,
      config.zkConnectionTimeoutMs, ZKStringSerializer)

  startWatchingTopicEvents()

  private def startWatchingTopicEvents() {
    val topicEventListener = new ZkTopicEventListener
    ZkUtils.makeSurePersistentPathExists(zkClient, ZkUtils.BrokerTopicsPath)

    zkClient.subscribeStateChanges(
      new ZkSessionExpireListener(topicEventListener))

    val topics = zkClient.subscribeChildChanges(
      ZkUtils.BrokerTopicsPath, topicEventListener).toList

    // call to bootstrap topic list
    topicEventListener.handleChildChange(ZkUtils.BrokerTopicsPath, topics)
  }

  private def stopWatchingTopicEvents() { zkClient.unsubscribeAll() }

  def shutdown() {
    lock.synchronized {
      try {
        if (zkClient != null) {
          stopWatchingTopicEvents()
          zkClient.close()
          zkClient = null
        }
        else
          logger.warn("Cannot shutdown already shutdown topic event watcher.")
      }
      catch {
        case e =>
          logger.fatal(e)
          logger.fatal(Utils.stackTrace(e))
      }
    }
  }

  class ZkTopicEventListener() extends IZkChildListener {

    @throws(classOf[Exception])
    def handleChildChange(parent: String, children: java.util.List[String]) {
      lock.synchronized {
        try {
          if (zkClient != null) {
            val latestTopics = zkClient.getChildren(ZkUtils.BrokerTopicsPath).toList
            logger.debug("all topics: %s".format(latestTopics))

            eventHandler.handleTopicEvent(latestTopics)
          }
        }
        catch {
          case e =>
            logger.fatal(e)
            logger.fatal(Utils.stackTrace(e))
        }
      }
    }

  }

  class ZkSessionExpireListener(val topicEventListener: ZkTopicEventListener)
    extends IZkStateListener {

    @throws(classOf[Exception])
    def handleStateChanged(state: KeeperState) { }

    @throws(classOf[Exception])
    def handleNewSession() {
      lock.synchronized {
        if (zkClient != null) {
          logger.info(
            "ZK expired: resubscribing topic event listener to topic registry")
          zkClient.subscribeChildChanges(
            ZkUtils.BrokerTopicsPath, topicEventListener)
        }
      }
    }
  }
}

File: kafka-0.7.0/kafka-0.7.0-core-java/src/main/java/kafka/consumer/ZookeeperTopicEventWatcher.java
package kafka.consumer;

import kafka.utils.Utils;
import kafka.utils.ZKStringSerializer;
import kafka.utils.ZkUtils;
import org.I0Itec.zkclient.IZkChildListener;
import org.I0Itec.zkclient.IZkStateListener;
import org.I0Itec.zkclient.ZkClient;
import org.apache.log4j.Logger;
import org.apache.zookeeper.Watcher;

import java.util.List;

public class ZookeeperTopicEventWatcher {
    private Logger logger = Logger.getLogger(getClass());
    private ConsumerConfig config;
    private TopicEventHandler<String> eventHandler;
    
    private Object lock = new Object();
    
    private ZkClient zkClient = new ZkClient(config.zkConnect, config.zkSessionTimeoutMs,
            config.zkConnectionTimeoutMs, new ZKStringSerializer());
    
    
    public ZookeeperTopicEventWatcher(ConsumerConfig config,
                                      TopicEventHandler<String> eventHandler) {
        this.config = config;
        this.eventHandler = eventHandler;
        startWatchingTopicEvents();
    }
    
    private void startWatchingTopicEvents() {
        ZkTopicEventListener topicEventListener = new ZkTopicEventListener();
        ZkUtils.makeSurePersistentPathExists(zkClient, ZkUtils.BrokerTopicsPath);
        
        zkClient.subscribeStateChanges(new ZkSessionExpireListener(topicEventListener));
    
        List<String> topics = zkClient.subscribeChildChanges(ZkUtils.BrokerTopicsPath, topicEventListener);
        
        // call to bootstrap topic list
        topicEventListener.handleChildChange(ZkUtils.BrokerTopicsPath, topics);
    }
    
    private void stopWatchingTopicEvents() { zkClient.unsubscribeAll(); }
    
    public void shutdown() {
         synchronized(lock) {
            try {
                if (zkClient != null) {
                    stopWatchingTopicEvents();
                    zkClient.close();
                    zkClient = null;
                }
                else
                    logger.warn("Cannot shutdown already shutdown topic event watcher.");
            }
            catch (Exception e){
                    logger.fatal(e);
                    logger.fatal(Utils.stackTrace(e));
            }
        }
    }
    
    class ZkTopicEventListener implements IZkChildListener {
        
        @Override
        public void handleChildChange(String parent, java.util.List<String> children) {
            synchronized (lock) {
                try {
                    if (zkClient != null) {
                        List<String> latestTopics = zkClient.getChildren(ZkUtils.BrokerTopicsPath);
                        logger.debug("all topics: %s".format(latestTopics.toArray().toString()));
                        eventHandler.handleTopicEvent(latestTopics);
                    }
                } catch (Exception e) {
                    logger.fatal(e);
                    logger.fatal(Utils.stackTrace(e));
                }
            }
        }
        
    }
    
    class ZkSessionExpireListener  implements IZkStateListener {
        ZkTopicEventListener topicEventListener;
        public ZkSessionExpireListener(ZkTopicEventListener topicEventListener) {
            this.topicEventListener = topicEventListener;
        }
    
        @Override
        public void handleStateChanged(Watcher.Event.KeeperState state) throws Exception {
        
        }
    
        @Override
        public void handleNewSession() throws Exception {
            synchronized(lock) {
                if (zkClient != null) {
                    logger.info("ZK expired: resubscribing topic event listener to topic registry");
                    zkClient.subscribeChildChanges(ZkUtils.BrokerTopicsPath, topicEventListener);
                }
            }
        }
    }
}
